File: app/src/main/java/com/android/calculator/model/ui/MainActivity.java
package com.android.calculator.model.ui;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import com.android.calculator.R;
import com.android.calculator.model.CalculatorImpl;
import com.android.calculator.model.Operator;

import java.util.HashMap;
import java.util.Map;

public class MainActivity extends AppCompatActivity implements CalculatorView {

    private TextView resultTxt;
    private CalculatorPresenter presenter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        resultTxt = findViewById(R.id.result);
        presenter = new CalculatorPresenter(this, new CalculatorImpl());
        Map<Integer, Integer> map = new HashMap<>();
        map.put(R.id.button_0, 0);
        map.put(R.id.button_1, 1);
        map.put(R.id.button_2, 2);
        map.put(R.id.button_3, 3);
        map.put(R.id.button_4, 4);
        map.put(R.id.button_5, 5);
        map.put(R.id.button_6, 6);
        map.put(R.id.button_7, 7);
        map.put(R.id.button_8, 8);
        map.put(R.id.button_9, 9);

        View.OnClickListener mapClickListener = new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                presenter.onPressed(map.get(view.getId()));
            }
        };

        findViewById(R.id.button_0).setOnClickListener(mapClickListener);
        findViewById(R.id.button_1).setOnClickListener(mapClickListener);
        findViewById(R.id.button_2).setOnClickListener(mapClickListener);
        findViewById(R.id.button_3).setOnClickListener(mapClickListener);
        findViewById(R.id.button_4).setOnClickListener(mapClickListener);
        findViewById(R.id.button_5).setOnClickListener(mapClickListener);
        findViewById(R.id.button_6).setOnClickListener(mapClickListener);
        findViewById(R.id.button_7).setOnClickListener(mapClickListener);
        findViewById(R.id.button_8).setOnClickListener(mapClickListener);
        findViewById(R.id.button_9).setOnClickListener(mapClickListener);

        Map<Integer, Operator> operatorsMap = new HashMap<>();
        operatorsMap.put(R.id.button_plus, Operator.ADD);
        operatorsMap.put(R.id.button_minus, Operator.SUB);
        operatorsMap.put(R.id.button_multi, Operator.MULTI);
        operatorsMap.put(R.id.button_div, Operator.DIV);

        View.OnClickListener operatorsMapClickListener = new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                presenter.onOperatorsPressed(operatorsMap.get(view.getId()));
            }
        };

        findViewById(R.id.button_plus).setOnClickListener(operatorsMapClickListener);
        findViewById(R.id.button_minus).setOnClickListener(operatorsMapClickListener);
        findViewById(R.id.button_multi).setOnClickListener(operatorsMapClickListener);
        findViewById(R.id.button_div).setOnClickListener(operatorsMapClickListener);

        findViewById(R.id.button_point).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                presenter.onPointPressed();
            }
        });
    }

    @Override
    public void showResult(String result) {
        resultTxt.setText(result);
    }

}

File: app/src/main/java/com/android/calculator/MainActivity.kt
package com.android.calculator

import android.annotation.SuppressLint
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.TextView
import com.android.calculator.viewmodel.CalculatorViewModel
import com.android.calculator.views.KeyboardKey
import com.android.calculator.views.KeyboardView

// Домашка. циклы,
//Функциональный тип, функция высшего порядка, паттерн listener
// ЖЦ активити знать.
// смена конфигурации.
// про inner / nested классы.
// про sealed классы.
// про typealias.
// ввод клавиш должен отображаться в expression_view.
//CE удаляет только последнее введённое число.

class MainActivity : AppCompatActivity() {

    private val viewModel = CalculatorViewModel()

    @SuppressLint("SetTextI18n")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val expressionView = findViewById<TextView>(R.id.expression_view)
        val resultView = findViewById<TextView>(R.id.result_view)

        // rv написать
        // 2-3 задачи на leetcode
        fun deleteSymbol() {
            if (expressionView.text.toString().isNotEmpty())
                expressionView.text = expressionView.text.toString().dropLast(1)
//                expressionView.text = expressionView.text.toString()
//                    .substring(0, expressionView.text.toString().length - 1)
            resultView.text = ""
        }

        findViewById<KeyboardView>(R.id.keyboard).setOnKeyPressedListener { key ->
            viewModel.onKeyPressed(key)

        }
    }

}

/*
InstrumentChartWidge
 when (key) {
                KeyboardKey.BUTTON_CE -> {
                    deleteSymbol()
                }

                KeyboardKey.BUTTON_MC -> {
                    expressionView.text = ""
                    resultView.text = ""
                }
                KeyboardKey.BUTTON_QUOTES -> expressionView.text =
                    expressionView.text.toString() + ">"
                KeyboardKey.BUTTON_DIV -> expressionView.text = expressionView.text.toString() + "/"
                KeyboardKey.BUTTON_7 -> expressionView.text = expressionView.text.toString() + "7"
                KeyboardKey.BUTTON_8 -> expressionView.text = expressionView.text.toString() + "8"
                KeyboardKey.BUTTON_9 -> expressionView.text = expressionView.text.toString() + "9"
                KeyboardKey.BUTTON_MULTI -> expressionView.text =
                    expressionView.text.toString() + "*"
                KeyboardKey.BUTTON_4 -> expressionView.text = expressionView.text.toString() + "4"
                KeyboardKey.BUTTON_5 -> expressionView.text = expressionView.text.toString() + "5"
                KeyboardKey.BUTTON_6 -> expressionView.text = expressionView.text.toString() + "6"
                KeyboardKey.BUTTON_MINUS -> expressionView.text =
                    expressionView.text.toString() + "-"
                KeyboardKey.BUTTON_1 -> expressionView.text = expressionView.text.toString() + "1"
                KeyboardKey.BUTTON_2 -> expressionView.text = expressionView.text.toString() + "2"
                KeyboardKey.BUTTON_3 -> expressionView.text = expressionView.text.toString() + "3"
                KeyboardKey.BUTTON_PLUS -> expressionView.text =
                    expressionView.text.toString() + "+"
                KeyboardKey.BUTTON_0 -> expressionView.text = expressionView.text.toString() + "0"
                KeyboardKey.BUTTON_COMMA -> expressionView.text =
                    expressionView.text.toString() + "."
                KeyboardKey.BUTTON_EQUALS -> expressionView.text =
                    expressionView.text.toString() + "="
            }
        }
* */
